Neos.ContentRepository.Core/Classes/ContentGraphFinder.php

```diff
@@ -0,0 +1,106 @@
+<?php
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+declare(strict_types=1);
+
+namespace Neos\ContentRepository\Core;
+
+use Neos\ContentRepository\Core\Projection\ContentGraph\ContentGraphInterface;
+use Neos\ContentRepository\Core\SharedModel\Exception\ContentStreamDoesNotExistYet;
+use Neos\ContentRepository\Core\SharedModel\Exception\WorkspaceDoesNotExist;
+use Neos\ContentRepository\Core\SharedModel\Workspace\ContentStreamId;
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+
+/**
+ * A finder for ContentGraphInterface bound to contentStream/Workspace
+ *
+ * @see ContentGraphInterface
+ * @api only for read access during write operations and in services
+ */
+final class ContentGraphFinder implements ContentGraphFinderInterface
```

Comment:
This thing here should get a nice explanation about the live cycle and cache behaviour of the ContentGraph to answer all questions ...

maybe to better explain caching and state the `ContentGraphFactoryInterface` should rather have a `getCurrentContentStreamIdForWorkspace` so the workspace name is manually resolved here and then put into `buildForWorkspaceAndContentStream`? That _could_ make things more obvious and that way we put all the heavy lifting here.

Reply:
Yes that could definitely be an option

Reply:
Some doc comments would surely help.
Nevertheless, I feel that we should lift the instance coordination to a higher level.
E.g. in Neos-land the new `Workspace` model will probably become the defacto standard way to access a content (sub)graph. So maybe it could keep the instances.

```php
foreach ($contentGraphFinder->getInstances() as $contentGraph) {
    foreach ($contentGraph->getSubgraphs() as $subgraph) {
        $subgraph->inMemoryCache->enable();
    }
}
```
This code is quite fragile (it's not new!) because it leaks a whole lot of internals (e.g. it relies on the `$subgraph` to have a field `inMemoryCache` etc).

IMO even a static `ContentSubgraphWithRuntimeCaches::resetRuntimeCaches()` or `ContentSubgraphWithRuntimeCaches::resetRuntimeCachesForInstance(ContentSubgraphIdentity $identity)` might work

Reply:
Yeah I saw that and didn't like it at all, but didn't spontaneously have a satisfactory idea. Not sure I like the static either.

Reply:
Kinda related to https://github.com/neos/neos-development-collection/pull/4955

if the behaviour dindt change, which i think it didnt, and we dont block any paths to a better future with this pr i would suggest to move this discussion to a separate thread to not block this pr.

Dont get me wrong, the runtime cache thing must be addressed but seeing this is a fundamental change for many prs to come i rather fix it later ;)

Reply:
> if the behaviour dindt change, which i think it didnt, and we dont block any paths to a better future with this pr
> i would suggest to move this discussion to a separate thread to not block this pr.

This adds public methods to userland code, that are not meant to be used by users. That's why it got slightly more important to get it right.

> this is a fundamental change for many prs to come i rather fix it later ;)

Absolutely! And as mentioned elsewhere: I don't mean to block this one, we can certainly iterate via follow-ups!

Reply:
We _could_ declare the CGFinder internal and relegate all user land cdoe to please only ever use CR->getContentGraph() to safeguard a bit for now, while we iterate 😆

Reply:
Ah a see sorry yes `ContentGraphFinder::reset` and `ContentGraphFinder::getInstances` are indeed new.
Previously we only exposed via the doctrine adapter `\Neos\ContentGraph\DoctrineDbalAdapter\...\ContentGraph::getSubgraphs` but we didnt have `public` methods declared inside of the core that are not meant to be used.
Declaring this thing internal would be my best bet too for now ;)